src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -3773,9 +3773,25 @@ long long QgsPostgresProvider::featureCount() const
     }
     else
     {
-      sql = QStringLiteral( "SELECT reltuples::bigint FROM pg_catalog.pg_class WHERE oid=regclass(%1)::oid" ).arg( quotedValue( mQuery ) );
+      sql = QStringLiteral( R"sql(
+        SELECT (CASE WHEN c.reltuples < 0 THEN NULL       -- never vacuumed
+                     WHEN c.relpages = 0 THEN float8 '0'  -- empty table
+                     ELSE c.reltuples / c.relpages END
+              * (pg_relation_size(c.oid) / pg_catalog.current_setting('block_size')::int)
+               )::bigint
+        FROM   pg_class c
+        WHERE  c.oid = regclass(%1)::oid;      -- schema-qualified table here
+        )sql" ).arg( quotedValue( mQuery ) );
       QgsPostgresResult result( connectionRO()->PQexec( sql ) );
       num = result.PQgetvalue( 0, 0 ).toLongLong();
+
+      // We've got 0 but it might be a table with a very small set of features, go for for exact count here
```

Comment:
Is there a risk that this might be a table with a very large set of features and it will slow things down a lot despite the use of estimated metadata being enabled?

Reply:
To my understanding the is such a risk if VACUUM or ANALYZE were never executed on the table.

Reply:
I'm happy to close this PR: I don't think that #46790 is really a bug: if you set the connection to use estimated metadata then you can expect that row count may be wrong.

Maybe we should better work on the GUI/UX side and make it clear to the user what are the consequences of setting the flag"use estimated metadata".

Reply:
@m-kuhn let's move the discussion to https://github.com/qgis/QGIS/issues/46790#issuecomment-1020053543

Reply:
Sounds all good